### PR TITLE
Feature: Fetch requests

### PR DESCRIPTION
**Pull request by bohdan-meltiukhov**
# Fetch requests
This PR adds fetch requests to the API service and the fetch-mock to catch these requests.

### CODE REVIEW

File: ui-components/app/services/api-service/index.js [hunk 1]
```diff
@@ -30,13 +83,45 @@ export default class ApiService {
...
+                errorCases.push(new ValidationErrorCase(error.field, error.message));
+              });
+              reject(new ServerValidationError(errorCases));
+            });
+            break;
+          case 500:
+            reject(new GeneralServerError('Internal server error'));
+            break;
+          default:
+            reject(new Error('Unknown error'));
```
Comment (dkalinin) [inline ui-components/app/services/api-service/index.js:110]: Seems that handling of server errors is similar in all cases. Let's create a factory for server errors.

File: ui-components/app/services/api-service/index.js [hunk 2]
```diff
@@ -2,41 +2,103 @@ import UserCredentials from '../../models/user-credentials';
...
+        error = new ServerValidationError(errorCases);
+      });
+      break;
+    case 500:
+      error = new GeneralServerError('Internal server error');
+      break;
+    default:
+      error = new Error('Unknown error');
+    }
+    return error;
```
Comment (dkalinin) [inline ui-components/app/services/api-service/index.js:90]: There's no need to have `error` variable. Just return errors in switch cases.

File: ui-components/tests/services/api-service/index.js [hunk 3]
```diff
@@ -1,36 +1,54 @@
...
 import UserCredentials from '../../../app/models/user-credentials';
+import fetchMock from '../../../node_modules/fetch-mock/esm/client.js';
+
 
 const {module, test} = QUnit;
 
-module('The ApiService test');
+module('The ApiService');
+
+test('should log in.', (assert) => {
```
Comment (dkalinin) [inline ui-components/tests/services/api-service/index.js:10]: Please add `assert.expect(3)` in this test.

File: ui-components/tests/services/api-service/index.js [hunk 4]
```diff
@@ -1,36 +1,54 @@
...
+  const password = '1234';
 
-test('should log in.', async function(assert) {
-  const apiService = new ApiService();
-  const userCredentials = new UserCredentials('admin', '1234');
+  fetchMock.post('/login', (url, opts) => {
+    const credentials = opts.body;
+    assert.strictEqual(credentials.username, username, 'The API Service should provide the correct username in the ' +
+      'request body.');
+    assert.strictEqual(credentials.password, password, 'The API Service should provide the correct password in the ' +
```
Comment (dkalinin) [inline ui-components/tests/services/api-service/index.js:18]: Mentioning of `The API Service` is redundant. We are in unit test of API Service.

File: ui-components/tests/services/api-service/index.js [hunk 5]
```diff
@@ -1,36 +1,54 @@
...
+  assert.ok(fetchMock.called('/login', {
+    method: 'POST',
+  }), 'The API Service should send the POST request to the \'/login\' URL.');
 });
 
-test('should register.', async function(assert) {
-  const apiService = new ApiService();
-  const userCredentials = new UserCredentials('user', 'password1234');
+test('should register.', (assert) => {
+  const username = 'admin';
```
Comment (dkalinin) [inline ui-components/tests/services/api-service/index.js:34]: Please add `assert.expect(3)` here.

File: ui-components/app/services/api-service/index.js (hunk 1 above)
Comment (bohdan-meltiukhov) [reply ui-components/app/services/api-service/index.js]: I created it now.

File: ui-components/app/services/api-service/index.js (hunk 2 above)
Comment (bohdan-meltiukhov) [reply ui-components/app/services/api-service/index.js]: I removed this unnecessary variable.

File: ui-components/tests/services/api-service/index.js (hunk 3 above)
Comment (bohdan-meltiukhov) [reply ui-components/tests/services/api-service/index.js:10]: Added.

File: ui-components/tests/services/api-service/index.js (hunk 5 above)
Comment (bohdan-meltiukhov) [reply ui-components/tests/services/api-service/index.js:38]: I added it now.

File: ui-components/tests/services/api-service/index.js (hunk 4 above)
Comment (bohdan-meltiukhov) [reply ui-components/tests/services/api-service/index.js]: I edited the assert messages.

File: ui-components/tests/services/api-service/index.js [hunk 6]
```diff
@@ -1,18 +1,27 @@
...
+import GeneralServerError from '../../../app/models/errors/general-server-error';
 
 const {module, test} = QUnit;
 
-module('The ApiService');
+const username = 'admin';
+const password = '1234';
+
+const apiService = ApiService.getInstance();
+const userCredentials = new UserCredentials(username, password);
```
Comment (dkalinin) [inline ui-components/tests/services/api-service/index.js]: Let's move this to test cases.

**Pull request merged**